### ISSUE TITLE
[BUG] cython package adds prerequisite on cython's internal files.

### ISSUE THREAD

**Issue by haampie** [cython/cython#5306]
### Describe the bug

When installing the following package: https://github.com/gevent/gevent/tree/1.5.0

I noticed that cython runs depending on the _mtime of the installation of cython_, which is frankly absurd. It looks like this:

```
Compiling src/gevent/resolver/cares.pyx because it depends on /opt/linux-ubuntu20.04-x86_64/gcc-11.1.0/py-cython-0.29.32-wistmswtv7bm55wpadctrr3eqb7jnona/lib/python3.10/site-packages/Cython/Includes/libc/string.pxd.
```

And I noticed this issue because we're setting mtime to 0 in binary tarballs, so that those tarballs are bitwise reproducible -- the mtime of cython became less than the mtime of the *sources* of gevent, causing cython *not to run*. 

This is pretty much the equivalent of setting a makefile target on a system library, like this: `target: source.c /usr/lib/libdependency.so`, which is a bad idea.

Is this something that can be fixed in cython?

Thanks

### OS

Linux

### Python version

3.10.8

### Cython version

0.29.33

**Comment by scoder** [comment]
> This is pretty much the equivalent of setting a makefile target on a system library, like this: `target: source.c /usr/lib/libdependency.so`, which is a bad idea.

It's rather like the C example `target: source.c /usr/include/string.h` which isn't quite that unreasonable. If you update your libc or your C compiler, then recompilation appears sensible to me. (I'm not saying that it has to be this way, but you might end up with random incompatibilities if you regenerate some C files based on the Cython version and not others, so it seems better to rebuild everything that depends on the updated C declarations that are part of Cython.)


> And I noticed this issue because we're setting mtime to 0 in binary tarballs, so that those tarballs are bitwise reproducible -- the mtime of cython became less than the mtime of the _sources_ of gevent, causing cython _not to run_.

If Cython does not run at all, then it means that the target C source file already exists. Why do you need Cython if that file has already been generated? If you really want to run Cython, then either a) pass the option `-f` or `force=True`, b)  do not include the generated C code in the source distribution to ensure a clean build, or c) use the `cython` command line tool instead of the `cythonize` tool.

I don't really see anything wrong here.

**Comment by haampie** [comment]
I disagree, even for local development, but that's besides the point. It's about releases and package management.

Tarball releases include source file mtimes, and they're typically restored when extracted. Nobody expects those files then to be regenerated depending on the mtime of a file _not in the tarball_. A package has a release tarball in 2020 with generated C sources included, you happen to have a Cython in your pythonpath, and now setup.py runs cythonize and regenerates those sources... That's obviously not the intention when folks create that release tarball.

> Why do you need Cython if that file has already been generated?

In the particular case it was a combination of two bugs: (1) the source files were generated with an old and buggy version of Cython, it did not work with Python 3.10, due to Cython relying too much on cpython internals. We never run into an issue though, because (2) cythonize erroneously regenerated C sources, because the release tarball mtimes were recorded as 2020 or w/e, and cython was installed as a build dep with recent mtime.

Now, in this case the very odd behavior of Cython was "helpful" in fixing a buggy release, but in general the exact opposite can happen too, where old and buggy Cython erroneously runs and produces a broken build.

Long story short: the Cython behavior right now is very unexpected, and totally unique to Cython; autotools packages don't do that for example. Please don't make build behavior depend on `mtime` of build deps (which isn't reliable across different systems anyways, only useful across files from the same release tarball generated on the same machine). Fixing this would help improve deterministic builds.

**Comment by scoder** [comment]
> cythonize erroneously regenerated C sources, because the release tarball mtimes were recorded as 2020 or w/e

Ok, but that makes it sound like a) you did not pin the exact Cython version and thus risked not getting reproducible results, and b) actually benefited from that. So, what exactly are you complaining about? *wink*

> Tarball releases include source file mtimes, and they're typically restored when extracted. Nobody expects those files then to be regenerated depending on the mtime of a file _not in the tarball_.

In any case, you seemed to be asking for reproducible builds. To that end, you need to pin the exact Cython release that you want to use for the build, and then don't have a problem, because Cython also (aims to) generate reproducible C files. Use the same version(s), get the same result. No problem here.

Note that you also need to pin the NumPy version, for example, if you make use of the NumPy-Cython interface declarations (`cimport numpy`). And the version of whatever compile time dependencies you are using (basically, all non-local `cimport`s). Basically, all compile time dependencies can have an impact on recompilation. And that is intended. When calling `cythonize()`, you are asking Cython to translate your sources into C/C++ code files. It can (sometimes) avoid doing that if it finds out that the target file is already there and its source file(s) and compile time dependencies did not change. But if they actually *did* change, then it cannot just sit there and do nothing. It has to do what you asked it to do.

What many users do instead, BTW, is to pin the Cython release series, and often do the same for the compile time dependencies. Thus, you benefit from improvements in the point releases (such as support for newer CPython versions), while keeping the risk of breakage low. But that's a user side decision and there are arguments for and against series- vs. exact version pinning.

I'm inclined to close this issue since I fail to see what we should improve. Excluding specifically the declarations that Cython ships doesn't feel right (although changes there should also get caught by an additional Cython version check in the target file that we do now, in Cython 3.0, I think), and excluding *all* compile time dependencies from the up-to-date check seems outright wrong. After all, the up-to-date check is mostly meant to avoid unnecessary costly recompilations, not to prevent changes in the target file.

**Comment by haampie** [comment]
> So, what exactly are you complaining about? *wink*

As I said, the reverse is possible: `setup.py` of a release tarball with generate C sources is run, a different cython *happens to be `import`able* (system install or whatever), and mtimes of completely incomparable things cause cython to regenerate C sources.

I don't know how to make clear to you that `mtime` between a release tarball and `mtime` of files of an installed packages are not comparable...

> But if they actually did change, then it cannot just sit there and do nothing. It has to do what you asked it to do.

Files in a release tarball as frozen, fixed. They shouldn't be regenerated based on whether you've installed cython before or after a random day in 2020. This is what folks rely on when releasing tarballs.

> and excluding all compile time dependencies from the up-to-date check seems outright wrong

The way I see it it's like `autotools`: there's no need to add `autoconf` / `automake` / `libtool` build deps when the release tarball contains the relevant generated configure scripts. However, if the mtimes of those files _in the release tarball_ are messed up, running `make` causes autoreconf etc to run (and will simply fail if build deps are missing).

What *never* happens is: `make` realizes that your system `automake`'s `mtime` is newer than some file *in the release tarball*, which would happen of course 99.9% of the time, and suddenly requires a full reconfigure. That's what's cython does and it's simply wrong.

**Comment by da-woods** [comment]
I'm inclined to agree with this issue, on the basis that it seems consistent with how system headers are treated in C. And that I'm not really sure we can reason too much about the validity of the timestamps of files outside the current project.

With that said, I also have little interest in trying to fix it, or in working out what exactly should be considered a system headers from a Cython point of view

**Comment by haampie** [comment]
For me it's not a high priority issue either, it's just that it bit me and it took a while to figure out what changed in our builds between one day and the next...
